### ISSUE TITLE
Empty Predictions when Applying Picard

### ISSUE THREAD

**Issue by atosystem** [ElementAI/picard#68]
Hi @tscholak,

Recently, I tried training T5-base for text2sql without using DB Schema and DB Content with your code.
The evaluation metrics seem normal without Picard.
When I apply Picard, I get plenty of empty predictions.
What might be the problem?

Thanks!

Myconfig:
```json
{
    "run_name": "t5-spider",
    "model_name_or_path": "/ckpts/train_spider_noSch/checkpoint-6656",
    "dataset": "spider",
    "source_prefix": "",
    "schema_serialization_type": "peteshaw",
    "schema_serialization_randomized": false,
    "schema_serialization_with_db_id": false,
    "schema_serialization_with_db_content": false,
    "normalize_query": true,
    "target_with_db_id": true,
    "output_dir": "/eval/eval_spider_noSch_nopicard/",
    "cache_dir": "/transformers_cache",
    "do_train": false,
    "do_eval": true,
    "fp16": false,
    "num_train_epochs": 3072,
    "per_device_train_batch_size": 16,
    "per_device_eval_batch_size": 1,
    "gradient_accumulation_steps": 16,
    "label_smoothing_factor": 0.0,
    "learning_rate": 1e-4,
    "adafactor": true,
    "adam_eps": 1e-6,
    "lr_scheduler_type": "constant",
    "warmup_ratio": 0.0,
    "warmup_steps": 0,
    "seed": 1,
    "report_to": [],
    "logging_strategy": "steps",
    "logging_first_step": true,
    "logging_steps": 16,
    "load_best_model_at_end": true,
    "metric_for_best_model": "exact_match",
    "greater_is_better": true,
    "save_total_limit": 128,
    "save_steps": 512,
    "evaluation_strategy": "steps",
    "eval_steps": 512,
    "predict_with_generate": true,
    "num_beams": 2,
    "num_beam_groups": 1,
    "use_picard": true,
    "launch_picard": true,
    "picard_mode": "parse_with_guards",
    "picard_schedule": "incremental",
    "picard_max_tokens_to_check": 2,
    "preprocessing_num_workers" : 32,
    "max_val_samples" : 1034
}
```

Empty Prediction Output:
```json
{
        "prediction": "",
        "query": "SELECT name ,  country ,  age FROM singer ORDER BY age DESC",
        "question": "What are the names, countries, and ages for every singer in descending order of age?",
        "context": "What are the names, countries, and ages for every singer in descending order of age? | stadium : stadium_id, location, name, capacity, highest, lowest, average | singer : singer_id, name, country, song_name, song_release_year, age, is_male | concert : concert_id, concert_name, theme, stadium_id, year | singer_in_concert : concert_id, singer_id",
        "label": "concert_singer | select name, country, age from singer order by age desc",
        "db_id": "concert_singer",
        "db_path": "/mydataset/spider/database",
        ...
```

Normal Prediction Output:
```json
{
        "prediction": "concert_singer | select name, country, age from singer order by age desc",
        "query": "SELECT name ,  country ,  age FROM singer ORDER BY age DESC",
        "question": "Show name, country, age for all singers ordered by age from the oldest to the youngest.",
        "context": "Show name, country, age for all singers ordered by age from the oldest to the youngest. | stadium : stadium_id, location, name, capacity, highest, lowest, average | singer : singer_id, name, country, song_name, song_release_year, age, is_male | concert : concert_id, concert_name, theme, stadium_id, year | singer_in_concert : concert_id, singer_id",
        "label": "concert_singer | select name, country, age from singer order by age desc",
        "db_id": "concert_singer",
        "db_path": "/mydataset/spider/database",
        ...
}
```

**Comment by tscholak** [comment]
Hi @atosystem,
Sorry to hear that. It appears to me that the source of the problem is `schema_serialization_with_db_id: False`. With that set to `False`, the model can't copy&paste the db id to the target, which is necessary for PICARD to function.

**Comment by atosystem** [comment]
Is there any workaround?
Perhaps give Picard the ground truth db id?

Also, another question, if my model fails to predict the right DB id (specifically, fail to copy&paste), then the Picard code will output nothing?

**Comment by tscholak** [comment]
Copy&pasting the db id is already a workaround. I chose that over hacking large portions of the huggingface code base. I'd gladly remove this hack.

If the model fails to reproduce the correct db id, then generation fails. However, PICARD also supervised the generation of db ids and constrains production to databases and schemas that have been registered with PICARD. Unless you have two schemas with extremely similar names, there shouldn't be an issue.

**Comment by atosystem** [comment]
Hi @tscholak,

Is it plausible that I insert the db_id tokenized array (ex: `concert_singer |`) after the `<pad>` token (which is the start token for t5 decoder) and before the rest of the generated token?

Specifically,
I override beam_search code for T5 model as the following:
Original:  ([code in huggingface](https://github.com/huggingface/transformers/blob/ef9c3ca348cb11c4ba951f8bdab068ecd8b847ce/src/transformers/generation_utils.py#L2111))
```python
next_tokens_scores = logits_processor(input_ids, next_token_logits)
``` 
Revised:
```python
next_tokens_scores = logits_processor(
                torch.cat(  
                    [
                        input_ids[:,0].unsqueeze(-1), 
                        model_kwargs["db_id"], 
                        input_ids[:,1:] 
                        ]  
                    , dim=-1)
                , next_token_logits)
```
where `model_kwargs["db_id"]` is the tokenized db_id concatenated with '|' and I only allow eval batch size to be 1

**Comment by tscholak** [comment]
This could work. If you want to reuse your own existing model, you should also change the decoder attention mask.

**Comment by atosystem** [comment]
@tscholak I'm not quite sure about the mention of the decoder attention mask.

Since I am only altering the inputs of the Picard logits processor, not the inputs of my decoder,  my decoder attention mechanism should work as usual?

**Comment by tscholak** [comment]
ah, that's right. perhaps this can work as is

**Comment by tscholak** [comment]
Hey @atosystem, I'm going to close this issue because I think there's nothing left I can help with. Feel free to reopen if you didn't get it working.
